src/api.ts

```diff
@@ -397,34 +408,94 @@ function createApi({
 
       const { port1, port2 } = new MessageChannel()
 
-      port2.on('message', (/** message: PortMessage */) => {
-        // TODO: do something with progress messages
-      })
+      activeImports.set(importId, port2)
+
+      return new Promise((res, rej) => {
+        // Initially use a longer duration to account for worker startup
+        let timeoutId = createTimeout(10000)
+
+        port2.on('message', handleFirstProgressMessage)
+        port2.on('message', resetTimeout)
+
+        // Can use a normal event emitter that emits an `abort` event as the abort signaler for Piscina,
+        // which allows us to not have to worry about globals or relying on polyfills
+        // https://github.com/piscinajs/piscina#cancelable-tasks
+        const abortSignaler = new EventEmitter()
+
+        piscina
+          .run(
+            {
+              dbPath: db.name,
+              importId,
+              mbTilesDbPath: mbTilesDb.name,
+              styleId,
+              tilesetId,
+              port: port1,
+            },
+            { signal: abortSignaler, transferList: [port1] }
+          )
+          .catch((err) => {
+            rej(err)
+          })
+          .finally(cleanup)
 
-      await piscina.run(
-        {
-          dbPath: db.name,
-          importId,
-          mbTilesDbPath: mbTilesDb.name,
-          styleId,
-          tilesetId,
-          port: port1,
-        },
-        { transferList: [port1] }
-      )
+        function handleFirstProgressMessage(message: PortMessage) {
+          if (message.type === 'progress') {
+            port2.off('message', handleFirstProgressMessage)
+            res({ import: { id: message.importId }, tileset })
+          }
+        }
+
+        function cleanup() {
+          clearTimeout(timeoutId)
+          port2.off('message', resetTimeout)
+          port2.close()
+          activeImports.delete(importId)
+        }
+
+        function onMessageTimeout() {
+          abortSignaler.emit('abort')
 
-      port2.close()
+          cleanup()
+
+          db.prepare<{
+            id: string
+            error: ImportError
+          }>(
+            "UPDATE Import SET state = 'error', finished = CURRENT_TIMESTAMP, error = :error WHERE id = :id"
+          ).run({ id: importId, error: 'TIMEOUT' })
+
+          rej(new Error('Timeout reached while waiting for worker message'))
+        }
```

Comment:
keeping an eye on this. i occasionally ran into cases locally where the db.prepare would throw an error because the db was closed. haven't reproduced for a while though...

Reply:
If this does throw an error will it be caught somewhere?

Reply:
i don't think so at the moment. does it make sense to wrap in a try/catch and reject on that caught error if it occurs?